code/ingest/src/main/java/org/jesterj/ingest/scanners/JdbcScanner.java

```diff
@@ -0,0 +1,295 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.jesterj.ingest.scanners;
+
+import java.sql.Connection;
+import java.sql.ResultSet;
+import java.sql.ResultSetMetaData;
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.jesterj.ingest.model.Document;
+import org.jesterj.ingest.model.Router;
+import org.jesterj.ingest.model.exception.ConfigurationException;
+import org.jesterj.ingest.model.exception.PersistenceException;
+import org.jesterj.ingest.model.impl.DocumentImpl;
+import org.jesterj.ingest.model.impl.ScannerImpl;
+import org.jesterj.ingest.utils.SqlUtils;
+
+import net.jini.space.JavaSpace;
+
+/**
+ * Scans a JDBC source such as an RDBMS (e.g. MySQL). Obtains a connection through the specified
+ * JDBC driver and extracts rows using the specified SQL statement; converts extracted rows to
+ * documents and passes the documents downstream.
+ * 
+ * @author dgoldenberg
+ */
+public class JdbcScanner extends ScannerImpl {
+
+  private static final Logger log = LogManager.getLogger();
+
+  private String jdbcDriver;
+  private String jdbcUrl;
+  private String jdbcUser;
+  private String jdbcPassword;
+
+  private String sqlStatement;
+
+  // TODO DG discuss
+  // May need to be set to Integer.MIN_VALUE for MySQL
+  // http://stackoverflow.com/questions/2447324/streaming-large-result-sets-with-mysql
+  // http://stackoverflow.com/questions/24098247/jdbc-select-batching-fetch-size-with-mysql
+  // http://stackoverflow.com/questions/20899977/what-and-when-should-i-specify-setfetchsize
+  private int fetchSize = -1;
+
+  private boolean autoCommit;
+  private int queryTimeout = -1;
+
+  @Override
+  public Function<String, String> getIdFunction() {
+    // TODO DG
+    return s -> s;
+  }
+
+  @Override
+  public Consumer<Document> getDocumentTracker() {
+    // TODO DG
+    return document -> {
+    };
+  }
+
+  @Override
+  public Runnable getScanOperation() {
+    return () -> {
+
+      try (
+        // Establish a connection and execute the query.
+        Connection conn = SqlUtils.createJdbcConnection(jdbcDriver, jdbcUrl, jdbcUser, jdbcPassword, autoCommit);
+
+        Statement statement = createStatement(conn);
+        ResultSet rs = statement.executeQuery(sqlStatement);) {
+        String[] columnNames = getColumnNames(rs);
+        int docIdColumnIdx = getDocIdColumnIndex(columnNames, getPlan().getDocIdField());
+
+        // TODO DG - what about the 'watch' thing / incrementals
```

Comment:
This is where getDocumentTracker needs to be enhanced. We should design an interface that can:

- record that a document was scanned and store time/hash
- answer the question when was this docid scanned last ( for comparison with mod time strategy)
- answer the question does this docid and hash match the stored hash for that document id. (for ignoring trivial mod time updates).